### PR TITLE
Added backward and forward compatibility integration tests for forkless upgrades

### PR DESCRIPTION
**Pull request by xgreenx**
Closes https://github.com/FuelLabs/fuel-core/issues/1622

The change adds integration tests for the forkless upgrade feature. The tests verify that it is backward compatible and, where possible, forward compatible.

The main idea is to use published `fuel-core` crates to produce blocks with different state transition functions.

Tests are using Ignition testnet chain config as a main source to be compatible with.

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: crates/services/importer/src/importer.rs [hunk 1]
```diff
@@ -204,7 +206,18 @@ where
...
+                tokio::time::timeout(tokio::time::Duration::from_secs(TIMEOUT), channel)
+                    .await;
+
+            if result.is_err() {
+                tracing::error!(
+                    "The previous block processing \
+                    was not finished for {TIMEOUT} seconds."
+                );
+                return Err(Error::PreviousBlockProcessingNotFinished)
+            }
```
Comment (xgreenx) [inline crates/services/importer/src/importer.rs:220]: Just added this to catch this problem in tests without debugging

File: crates/services/p2p/src/service.rs [hunk 2]
```diff
@@ -89,8 +89,6 @@ pub type Service<V> = ServiceRunner<UninitializedTask<V, SharedState>>;
 enum TaskRequest {
     // Broadcast requests to p2p network
     BroadcastTransaction(Arc<Transaction>),
-    // Request to get one-off data from p2p network
-    GetPeerIds(oneshot::Sender<Vec<PeerId>>),
```
Comment (xgreenx) [inline crates/services/p2p/src/service.rs:93]: It is not used, so I removed it=)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -122,8 +122,8 @@ jobs:
             args: -p fuel-core-client --no-default-features
           - command: test
             args: -p fuel-core-chain-config --no-default-features
-          - command: test
-            args: --manifest-path version-compatibility/Cargo.toml --workspace
+          - command: check
+            args: --manifest-path version-compatibility/Cargo.toml --workspace && cargo test --manifest-path version-compatibility/Cargo.toml --workspace
```
Comment (MitchTurner) [inline .github/workflows/ci.yml:126]: Curious why you stuck this into a single command? Feels weird calling `cargo` in part of the `args`.

Comment (MitchTurner) [inline .github/workflows/ci.yml:126]: Curious why you stuck this into a single command? Feels weird calling `cargo` in part of the `args`.

File: version-compatibility/forkless-upgrade/README.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Forkless upgrade tests
+
+This crate tests that state transition functions for all releases of `fuel-core` are backward compatible. 
+
+In addition, we also test that releases are forward-compatible unless we introduce a breaking change in the API.
+
+## Adding new test
+
+We need to add a new backward compatibility test for each new release. To add tests, we need to duplicate tests that are using the latest `fuel-core` and replace usage of the latest crate with a new release.
```
Comment (MitchTurner) [inline version-compatibility/forkless-upgrade/README.md:9]: Is there anything enforcing that we do this?

File: version-compatibility/forkless-upgrade/src/backward_compatibility.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        "--peering-port",
+        genesis_port,
+    ])
+    .await
+    .unwrap();
+    let public_key = Keypair::from(genesis_keypair).public();
+    let genesis_peer_id = PeerId::from_public_key(&public_key);
+    let genesis_multiaddr = default_multiaddr(genesis_port, genesis_peer_id);
+
+    // Starting node that uses latest furl core.
```
Comment (MitchTurner) [inline version-compatibility/forkless-upgrade/src/backward_compatibility.rs:68]:
```suggestion
    // Starting node that uses latest fuel core.
```

File: version-compatibility/forkless-upgrade/README.md (hunk 4 above)
Comment (xgreenx) [reply version-compatibility/forkless-upgrade/README.md:9]: Nope, and I don't know who to enforce it=)

File: .github/workflows/ci.yml (hunk 3 above)
Comment (xgreenx) [reply .github/workflows/ci.yml:126]: Added a comment

File: version-compatibility/forkless-upgrade/src/backward_compatibility.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    ])
+    .await;
+
+    // Then
+    let latest_node = latest_node.expect("Failed to spawn latest node");
+    assert!(latest_node.node.state().started())
+}
+
+#[tokio::test]
+async fn latest_binary_follows_blocks_created_by_genesis_binary() {
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply version-compatibility/forkless-upgrade/src/backward_compatibility.rs:41]: The module itself is named `backward_compatibility`; I thought it should be clear already=D Added information about backward compatibility.

File: version-compatibility/forkless-upgrade/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#![deny(unused_crate_dependencies)]
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply version-compatibility/forkless-upgrade/src/lib.rs:1]: Why we don't need it?=) It is to be sure that all dependencies are used.

File: version-compatibility/forkless-upgrade/src/forward_compatibility.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    PeerId,
+};
+use rand::{
+    rngs::StdRng,
+    SeedableRng,
+};
+use std::time::Duration;
+
+#[tokio::test]
+async fn latest_state_transition_function_is_forward_compatible_with_genesis_binary() {
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply version-compatibility/forkless-upgrade/src/forward_compatibility.rs:34]: It already contains this information 
image

File: crates/services/p2p/src/service.rs (hunk 2 above)
Comment (Voxelot) [reply crates/services/p2p/src/service.rs:93]: it would be cleaner to make this kind of change in a separate PR, as it's totally unrelated to backwards compatibility right?

File: version-compatibility/forkless-upgrade/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MitchTurner) [reply version-compatibility/forkless-upgrade/src/lib.rs:1]: Hmm. Idk what I was seeing when I wrote this. Must have read it as `allow`.

**Review by MitchTurner — APPROVED** [review]

File: crates/services/p2p/src/service.rs (hunk 2 above)
Comment (xgreenx) [reply crates/services/p2p/src/service.rs:93]: https://github.com/FuelLabs/fuel-core/pull/1913
